### ISSUE TITLE
`useLazyQuery` execute function never resolves if component is immediately unmounted

### ISSUE THREAD

**Issue by a-churchill** [apollographql/apollo-client#10174]
**Intended outcome:**

Calling the `execute` function returned by `useLazyQuery` always resolves with the result of the query.

**Actual outcome:**

The `execute` function never resolves, and no network request is ever made.

**How to reproduce the issue:**

[Here is a CodeSandbox](https://codesandbox.io/s/apollo-client-uselazyquery-bug-r4kzvi) that reproduces the issue. It is a toy example that mimics the exact circumstances this bug occurred in real application code (which I can't share publicly). The console logs show the `execute` function is called but never resolves: 
<img width="700" alt="image" src="https://user-images.githubusercontent.com/18105223/195177374-10d8853d-c941-4028-b415-90172d6cbac4.png">

**Versions**
The issue is reproduced in a CodeSandbox with React 17.0.2 and Apollo 3.6.8. Here is the output for my machine too, if relevant:

```
$ npx envinfo@latest --preset apollo --clipboard

  System:
    OS: macOS 12.6
  Binaries:
    Node: 16.17.1 - ~/.nvm/versions/node/v16.17.1/bin/node
    Yarn: 3.1.0-rc.7.git.20210926.hash-f0d86bd4 - ~/.nvm/versions/node/v16.17.1/bin/yarn
    npm: 8.15.0 - ~/.nvm/versions/node/v16.17.1/bin/npm
  Browsers:
    Chrome: 106.0.5249.103
    Safari: 16.0
```

**Comment by jpvajda** [comment]
@a-churchill Thanks for letting us know about this and the great reproduction. for fun I upgraded this sandbox to our newest 3.7.0 version, and I'm seeing the same problem. 



<img width="1369" alt="Screen Shot 2022-10-11 at 5 53 01 PM" src="https://user-images.githubusercontent.com/1395158/195464032-b542594e-1a86-4d71-808f-05b9da8ce8c9.png">

**Comment by jerelmiller** [comment]
Hey @a-churchill 👋 

We just released [v3.7.11](https://github.com/apollographql/apollo-client/releases/tag/v3.7.11) which fixes this issue. `useLazyQuery` will now ensure that promises that are not resolved before a component unmounts will allow that promise to resolve naturally with data from the request. I tried this out in a [custom fork of your reproduction](https://codesandbox.io/s/apollo-client-uselazyquery-bug-forked-99v7hy) that demonstrates the correct behavior. Thanks for reporting the issue!
